test/funnels/Swapper.t.sol

```diff
@@ -0,0 +1,146 @@
+
+    address constant FACILITATOR = address(0x1337);
+    address constant KEEPER      = address(0xb0b);
+
+    uint8 constant SWAPPER_ROLE = uint8(1);
+
+    function setUp() public {
+        vm.createSelectFork(vm.envString("ETH_RPC_URL"));
+
+        buffer = new AllocatorBuffer(ilk);
+        roles = new AllocatorRoles();
+        swapper = new Swapper(address(roles), ilk, address(buffer));
+        uniV3Callee = new UniV3SwapperCallee(UNIV3_ROUTER);
+
+        roles.setIlkAdmin(ilk, address(this));
+        roles.setRoleAction(ilk, SWAPPER_ROLE, address(swapper), swapper.swap.selector, true);
+        roles.setUserRole(ilk, FACILITATOR, SWAPPER_ROLE, true);
+
+        swapper.file("cap", DAI, USDC, 10_000 * WAD);
+        swapper.file("cap", USDC, DAI, 10_000 * 10**6);
+        swapper.file("hop", DAI, USDC, 3600);
+        swapper.file("hop", USDC, DAI, 3600);
+
+        deal(DAI,  address(buffer), 1_000_000 * WAD,   true);
+        deal(USDC, address(buffer), 1_000_000 * 10**6, true);
+        buffer.approve(USDC, address(swapper), type(uint256).max);
+        buffer.approve(DAI,  address(swapper), type(uint256).max);
+    }
+
+    function testConstructor() public {
+        Swapper s = new Swapper(address(0xBEEF), "SubDAO 1", address(0xAAA));
+        assertEq(address(s.roles()),  address(0xBEEF));
+        assertEq(s.ilk(), "SubDAO 1");
+        assertEq(s.wards(address(this)), 1);
+    }
+
+    function testAuth() public {
+        checkAuth(address(swapper), "Swapper");
+    }
+
+    function testModifiers() public {
+        bytes4[] memory authedMethods = new bytes4[](1);
+        authedMethods[0] = swapper.swap.selector;
+
+        vm.startPrank(address(0xBEEF));
+        checkModifier(address(swapper), "Swapper/not-authorized", authedMethods);
+        vm.stopPrank();
+    }
+
+    function testFile() public {
+        checkFileUintForGemPair(address(swapper), "Swapper", ["cap", "hop"]);
+    }
+
+    function testRoles() public {
+        vm.expectRevert("Swapper/not-authorized");
+        vm.prank(address(0xBEEF)); swapper.file("hop", address(0), address(0), 0);
+        roles.setRoleAction(ilk, uint8(0xF1), address(swapper), bytes4(keccak256("file(bytes32,address,address,uint256)")), true);
+        roles.setUserRole(ilk, address(0xBEEF), uint8(0xF1), true);
+        vm.prank(address(0xBEEF)); swapper.file("hop", address(0), address(0), 0);
+    }
+
+    function testSwap() public {
+        uint256 prevDst = GemLike(DAI).balanceOf(address(buffer));
+
+        vm.expectEmit(true, true, true, false);
+        emit Swap(FACILITATOR, USDC, DAI, 10_000 * 10**6, 0);
+        vm.prank(FACILITATOR); uint256 out = swapper.swap(USDC, DAI, 10_000 * 10**6, 9900 * WAD, address(uniV3Callee), USDC_DAI_PATH);
+        
+        assertGe(out, 9900 * WAD);
+        assertEq(GemLike(DAI).balanceOf(address(buffer)), prevDst + out);
+        assertEq(GemLike(DAI).balanceOf(address(swapper)), 0);
```

Comment:
Maybe we should also check that the amount of USDC in the buffer went down as expected.

Reply:
Implemented at https://github.com/makerdao/dss-allocator/pull/14.